### ISSUE TITLE
BUG: order of promoters matters

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#26104]
As noted by @mhvk [here](https://github.com/numpy/numpy/pull/26102#discussion_r1534636116), the order that promoters get added matters, and it shouldn't.

Taking a look at the promoter resolution algorithm:

https://github.com/numpy/numpy/blob/6bac54efcb5a4b6b099607f6b4372b5f20e3e715/numpy/_core/src/umath/dispatching.c#L244-L471

Indeed, it does exit if it finds two possible ambiguous promoters and doesn't check any subsequent ones for better matches.

**Comment by seberg** [comment]
Right, NumPy would have to keep track of all "best" candidates hoping that a later one will be better than all previous ones.
That seemed like unnecessary complexity at the time (and might even still be now).

So right now the more interesting/urgent thing is to add the abstract DTypes (even if they must only be used for registering promotion!).  But of course that doesn't mean that it may not be good to fix this issue!

**Comment by mhvk** [comment]
Hopefully with #26116 we don't need to worry about this any more...

**Comment by ngoldbaum** [comment]
Agreed that eliminates a lot of the need for generic promoters and the need for running this algorithm, but there still might be cases where the algorithm still should check for a better match "after" one its already found. We may hit other limitations as we use the promoters more elsewhere in numpy or in other packages.
